### PR TITLE
[Wallet] get rid of pwalletMain, add simple CWallets infrastructure

### PR DESCRIPTION
**Pull request by jonasschnelli**
This is a very simple pull requests that removes the global `pwalletMain` pointer and replaces the calls with `CWallets::defaultWallet()`.

This PR leverages the internal global single wallet container to a vector. We have similar functionality on the Qt layer.

This will allow a step-by-step transition towards multiwallet support like proposed by @luke-jr in #8694.

### CODE REVIEW

File: src/init.cpp [hunk 1]
```diff
@@ -231,8 +230,8 @@ void Shutdown()
         pblocktree = NULL;
     }
 #ifdef ENABLE_WALLET
-    if (pwalletMain)
-        pwalletMain->Flush(true);
+    if (CWallets::defaultWallet())
```
Comment (laanwj) [inline src/init.cpp]: CWallets::FlushAllWallets() ?

File: src/wallet/wallet.h [hunk 2]
```diff
@@ -986,4 +984,37 @@ class CAccount
...
+        LOCK(cs_wallets);
+        if (!wallets.size())
+            return NULL;
+        return wallets[0];
+    }
+    static void addWallet(CWallet *newWallet)
+    {
+        LOCK(cs_wallets);
+        wallets.push_back(newWallet);
+        int i = 0;
```
Comment (paveljanik) [inline src/wallet/wallet.h]: `i` looks unused here 8)

Comment (jonasschnelli) [inline src/wallet/wallet.h]: That's black magic! :-)

Will remove it, had it there for debug purposes.

File: src/init.cpp (hunk 1 above)
Comment (jonasschnelli) [inline src/init.cpp]: Yes. Changed it to `flushAllWallets(bool shutdown);`

File: src/init.cpp [hunk 3]
```diff
@@ -1535,9 +1531,9 @@ bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
     uiInterface.InitMessage(_("Done loading"));
 
 #ifdef ENABLE_WALLET
-    if (pwalletMain) {
+    if (CWallets::defaultWallet()) {
         // Run a thread to flush wallet periodically
-        threadGroup.create_thread(boost::bind(&ThreadFlushWalletDB, boost::ref(pwalletMain->strWalletFile)));
+        threadGroup.create_thread(boost::bind(&ThreadFlushWalletDB, boost::ref(CWallets::defaultWallet()->strWalletFile)));
```
Comment (jonasschnelli) [inline src/init.cpp]: We keep the `ThreadFlushWalletDB` with "capacity" for a single wallet file for now.

File: src/wallet/wallet.h [hunk 4]
```diff
@@ -986,4 +984,36 @@ class CAccount
     }
 };
 
+static CCriticalSection cs_wallets;
+class CWallets
+{
+public:
+    static std::vector<CWallet *> wallets;
```
Comment (MarcoFalke) [inline src/wallet/wallet.h]: Why public?

File: src/qt/signverifymessagedialog.cpp [hunk 5]
```diff
@@ -142,7 +142,7 @@ void SignVerifyMessageDialog::on_signMessageButton_SM_clicked()
     }
 
     CKey key;
-    if (!pwalletMain->GetKey(keyID, key))
+    if (!CWallets::defaultWallet()->GetKey(keyID, key))
```
Comment (paveljanik) [inline src/qt/signverifymessagedialog.cpp]: There is no `if (pwalletMain)` test here in the original code?

Comment (MarcoFalke) [inline src/qt/signverifymessagedialog.cpp]: @paveljanik The dialog does not exist when there is no wallet.

File: src/init.cpp (hunk 3 above)
Comment (MarcoFalke) [inline src/init.cpp]: Why?

File: src/wallet/rpcdump.cpp [hunk 6]
```diff
@@ -270,7 +270,7 @@ UniValue importprunedfunds(const UniValue& params, bool fHelp)
     if (!DecodeHexTx(tx, params[0].get_str()))
         throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed");
     uint256 hashTx = tx.GetHash();
-    CWalletTx wtx(pwalletMain,tx);
+    CWalletTx wtx(CWallets::defaultWallet(),tx);
```
Comment (paveljanik) [inline src/wallet/rpcdump.cpp]: SPC after `,` :-)

**Comment by paveljanik** [comment]
I'm all for supporting multiple wallets, so concept ACK.

But isn't this only the part of @luke-jr 's work (64bcde7c5535cda12fc20ded718558ca74b28191)? Can you please coordinate the work to save your time? ;-)

File: src/wallet/rpcdump.cpp (hunk 6 above)
Comment (MarcoFalke) [inline src/wallet/rpcdump.cpp]: Or just `git diff HEAD~ -U0|cfd`.

```sh
alias cfd='~/workspace/bitcoin/contrib/devtools/clang-format-diff.py -p1 -i -v'

**Comment by jonasschnelli** [comment]
I really like @luke-jr https://github.com/bitcoin/bitcoin/pull/8694, though, my experience with lager pull requests tells me, that #8694 will have a hard time getting in as a whole. I try to slice out independent chunks, improve them and PR them separately. This is how I get you all reviewing parts of #8694.

File: src/wallet/wallet.h (hunk 4 above)
Comment (jonasschnelli) [inline src/wallet/wallet.h]: Yes. I had a `Namespace` at the beginning and though that there is no big need for a Class or even private static class variables. But yes, why not private. Will update.

File: src/init.cpp (hunk 3 above)
Comment (jonasschnelli) [inline src/init.cpp]: 1. Scope creeping
2. Moving target (`ThreadFlushWalletDB` is currently getting touches in different PRs).

**Comment by jonasschnelli** [comment]
Fixed nits.

File: src/init.cpp (hunk 3 above)
Comment (MarcoFalke) [inline src/init.cpp]: Fine, can be done later. Though, I think in case it is just a simple for loop, it should be part of this pull.

Comment (jonasschnelli) [inline src/init.cpp]: I think we don't want a flush wallet thread per wallet. IMO it should be a single flush thread that loops though the wallet, check last-flush-time and optionally flushes.

Comment (MarcoFalke) [inline src/init.cpp]: Jup, @luke-jr did this. (Create a loop in `ThreadFlushWalletDB` over all wallets)

File: src/wallet/wallet.h [hunk 7]
```diff
@@ -986,4 +984,37 @@ class CAccount
...
+    static void flushAllWallets(bool shutdown)
+    {
+        LOCK(cs_wallets);
+        for(CWallet *pWallet : wallets)
+            pWallet->Flush(shutdown);
+    }
+    static void destruct()
+    {
+        LOCK(cs_wallets);
+        for(CWallet *pWallet : wallets)
```
Comment (MarcoFalke) [inline src/wallet/wallet.h]: μNit

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 656d344..d690e48 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -61 +61 @@ const uint256 CMerkleTx::ABANDON_HASH(uint256S("00000000000000000000000000000000
-std::vector<CWallet *> CWallets::wallets = {};
+std::vector<CWallet*> CWallets::wallets = {};
diff --git a/src/wallet/wallet.h b/src/wallet/wallet.h
index 390210d..4fb69dd 100644
--- a/src/wallet/wallet.h
+++ b/src/wallet/wallet.h
@@ -991 +991,2 @@ private:
-    static std::vector<CWallet *> wallets;
+    static std::vector<CWallet*> wallets;
+
@@ -1000 +1001 @@ public:
-    static void addWallet(CWallet *newWallet)
+    static void addWallet(CWallet* newWallet)
@@ -1008 +1009 @@ public:
-        for(CWallet *pWallet : wallets)
+        for (CWallet* pWallet : wallets)
@@ -1014 +1015 @@ public:
-        for(CWallet *pWallet : wallets)
+        for (CWallet* pWallet : wallets)

**Review by maflcko — APPROVED** [review]
utACK 9cc96b59f4bcf708ae370679ef41c6f8f2af3a14

File: src/wallet/wallet.h [hunk 8]
```diff
@@ -986,4 +984,37 @@ class CAccount
...
+    static void addWallet(CWallet *newWallet)
+    {
+        LOCK(cs_wallets);
+        wallets.push_back(newWallet);
+    }
+    static void flushAllWallets(bool shutdown)
+    {
+        LOCK(cs_wallets);
+        for(CWallet *pWallet : wallets)
+            pWallet->Flush(shutdown);
```
Comment (luke-jr) [inline src/wallet/wallet.h]: Does cs_wallets need to be held for access to (flushing) the wallet itself, or only for the wallets vector? If the former, then defaultWallet will fail to guard it; if the latter, the lock here is unnecessarily long-held.

**Comment by luke-jr** [comment]
#8694 is already step-by-step, so this mostly just forces rebasing it... IMO better to do refactoring after #8694 gets merged. :)

File: src/wallet/wallet.h (hunk 8 above)
Comment (jonasschnelli) [inline src/wallet/wallet.h]: I decided to hold the lock because it would be possible that during iteration a addWallet (or later removeWallet) happens. But right,.. it results in holding the lock for probably a long time.

I first though doing a copy of the wallets vector would be good, but the CWallet instance where the copy point to might be gone.

Maybe we keep this for now because the flushAllWallets happens only on shutdown.
Or what would be a better solution?

**Comment by jonasschnelli** [comment]
Rebased and fixed code formatting (@MarcoFalkes micro-nit).
The diff is relatively large, but it would finally eliminate the global `pwalletMain` in favor of a more flexible handler that would be "ready" for multiwallet.

**Comment by MarcoFalke** [comment]
Needs rebase

**Comment by jonasschnelli** [comment]
Rebased (now also covers `importmulti`).

**Comment by ryanofsky** [comment]
Seems like currently the first multiwallet commit is #8775, then #8694 comes next, and maybe this will be rebased on top of those at a later point?

**Comment by jonasschnelli** [comment]
Closing. Since #8775, this is to far away from being useful. Maybe a different similar approach can be worked out.
